crates/ursa-rpc-service/src/api.rs

```diff
@@ -283,6 +241,177 @@ where
     }
 }
 
+impl<S> NodeNetworkInterface<S>
+where
+    S: Blockstore + Store + Send + Sync + 'static,
+{
+    pub fn new(
+        store: Arc<UrsaStore<S>>,
+        network_send: Sender<NetworkCommand>,
+        provider_send: Sender<ProviderCommand>,
+        origin_config: OriginConfig,
+    ) -> Self {
+        Self {
+            store,
+            network_send,
+            provider_send,
+            origin_config,
+            pending_requests: Arc::new(RwLock::new(HashMap::new())),
+            client: Arc::new(Client::new()),
+        }
+    }
+
+    /// Ensure a root cid is synced to the blockstore
+    async fn sync_content(&self, cid: Cid) -> Result<()> {
+        if !self.store.blockstore().has(&cid)? {
+            info!("Requesting block with the cid {cid:?}");
+            if let Err(e) = self.get_network(cid).await {
+                info!("Failed to get content from network: {}", e);
+                let size = self.get_origin(cid).await?;
+                self.provide_cid(cid, Some(size)).await
+            } else {
+                self.provide_cid(cid, None).await
```

Comment:
isn't `size` required for the announcement?

Reply:
If size is None, then `provide_cid` will construct a temporary file and get the length using store.car_size(cid)

Reply:
I don't follow, bitswap relationship with announcement size

Reply:
Size shouldn't be None

Reply:
With origin, we get back a car response body, so we just .len() and done

With bitswap, fetched content is block by block and doesn't return a car size to use. So provide_cid accepts either a given size, or None which will make it calculate the car size on its own.

https://github.com/fleek-network/ursa/blob/841b87224aeeea04dd1e7a160475a48eca23b0e0/crates/ursa-rpc-service/src/api.rs#L380-L383

Reply:
I am not sure you are following the comment I made, the size should never be None right. The function should just provide the cid with the size. How it gets is addressed in another comment

Reply:
ah okay, I can move the call to car_size and make provide_cid always accept a size

Reply:
https://github.com/fleek-network/ursa/pull/278/commits/171e54140367661a9e63e432d178b3ed9b0f876a